### PR TITLE
[SPARK-31231][BUILD] Explicitly setuptools version as 46.0.0 in pip package test

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

For a bit of background,
PIP packaging test started to fail (see [this logs](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120218/testReport/)) as of  setuptools 46.1.0 release. In https://github.com/pypa/setuptools/issues/1424, they decided to don't keep the modes in `package_data`.

In PySpark pip installation, we keep the executable scripts in `package_data` https://github.com/apache/spark/blob/fc4e56a54c15e20baf085e6061d3d83f5ce1185d/python/setup.py#L199-L200, and expose their symbolic links as executable scripts.

So, the symbolic links (or copied scripts) executes the scripts copied from `package_data`, which doesn't have the executable permission in its mode:

```
/tmp/tmp.UmkEGNFdKF/3.6/bin/spark-submit: line 27: /tmp/tmp.UmkEGNFdKF/3.6/lib/python3.6/site-packages/pyspark/bin/spark-class: Permission denied
/tmp/tmp.UmkEGNFdKF/3.6/bin/spark-submit: line 27: exec: /tmp/tmp.UmkEGNFdKF/3.6/lib/python3.6/site-packages/pyspark/bin/spark-class: cannot execute: Permission denied
```

The current issue is being tracked at https://github.com/pypa/setuptools/issues/2041

</br>

For what this PR proposes:
It sets the setuptools version in PR builder for now to unblock other PRs.  _This PR does not solve the issue yet_. I will make a fix after monitoring https://github.com/pypa/setuptools/issues/2041

### Why are the changes needed?

It currently affects users who uses the latest setuptools. So, _users seem unable to use PySpark with the latest setuptools._ See also https://github.com/pypa/setuptools/issues/2041#issuecomment-602566667

### Does this PR introduce any user-facing change?

It makes CI pass for now. No user-facing change yet.

### How was this patch tested?

Jenkins will test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Seems a bug from the latest `setuptools`.

**Comment by HyukjinKwon** [comment]
Damn, okay. it was because of new setuptool release, specially this https://github.com/pypa/setuptools/commit/e7f29530df7595a876e4995f2011784ac9f43086.

**Comment by HyukjinKwon** [comment]
Okay .. it passed. I'll clean up and merge to unblock other PRs.

File: dev/run-pip-tests [hunk 1]
```diff
@@ -110,6 +110,7 @@ for python in "${PYTHON_EXECS[@]}"; do
     cd /
 
     echo "Run basic sanity check on pip installed version with spark-submit"
+    # ls -al $(dirname $(dirname $(which spark-submit)))/lib/python3.6/site-packages/pyspark/bin/spark-class
```
Comment (cloud-fan) [inline dev/run-pip-tests]: tests pass now, shall we remove these commented code?

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.0, and branch-2.4.

I am not going to resolve the JIRA yet - this PR is a temp workaround at this momoent.

**Comment by HyukjinKwon** [comment]
@nchammas, @srowen, @dongjoon-hyun, @BryanCutler, FWIW, I think this is one example of the trade-off we discussed at https://github.com/apache/spark/pull/27928.

`setuptools` is dev-only dependency but seems it found a pretty critical issue. Currently seems users can't use pip-installed PySpark assuming from pypa/setuptools#2041. However, of course it broke the PR builders.

**Comment by srowen** [comment]
Was there an issue in 46.1.0? wouldn't pinning to 46.0.0 have been a good thing then? or else I misunderstand what broke.

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon .
This broke branch-2.4 due to the following. I'll revert this from `branch-2.4`.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-2.4-test-sbt-hadoop-2.6/804/
```
UnsatisfiableError: The following specifications were found to be in conflict:
  - python=3.5
  - setuptools=46.0.0 -> python[version='>=3.8,<3.9.0a0'] -> openssl[version='>=1.1.1d,<1.1.2a']
Use "conda search <package> --info" to see the dependencies for each package.
```

**Comment by dongjoon-hyun** [comment]
Since Apache Spark 2.4 still supports `Python 2.7+/3.4+`, `setuptools 46.0.0` seems to be incompatible to that. Please make a separate PR for `branch-2.4`.

**Comment by nchammas** [comment]
Perhaps the requirement should instead be of the form `setuptools < 46.1.0`, to allow flexibility for older versions of Python to still find a release of setuptools that satisfies their requirements.

**Comment by HyukjinKwon** [comment]
Yeah, maybe I should better set upperbound instead. Thanks @dongjoon-hyun and @nchammas

**Comment by HyukjinKwon** [comment]
@srowen, seems this is the problem in client side with `setuptools`.So, using lower version now fixed the test but the issue itself persists.

**Comment by srowen** [comment]
I'm still confused. This is an argument for _pinning_ versions, right? Because a new version of `setuptools` caused a problem:

**Comment by HyukjinKwon** [comment]
But we found that issue from the broken PR builder, which seems to be pretty critical.

**Comment by HyukjinKwon** [comment]
I simply meant this is an example of the trade-off. It caused a problem and broke the PR builder vs it found a problem early.

**Pull request closed without merging**